### PR TITLE
Update API tests for hosts

### PR DESCRIPTION
**Pull request by Ichimonji10**
Update module `tests.foreman.api.test_host`:

* Drop DDT. (See #2769.)
* Make update test re-use a single host. This speeds up the test.

Test results:

    $ nosetests tests/foreman/api/test_host.py | less
    ....
    ----------------------------------------------------------------------
    Ran 4 tests in 18.180s

    OK

### CODE REVIEW

**Comment by Ichimonji10** [comment]
Rebased.

File: tests/foreman/api/test_host.py [hunk 1]
```diff
@@ -70,27 +63,29 @@ def test_create_owner_type(self, owner_type):
...
-        host = host.create(create_missing=False)
-        self.assertEqual(host.owner_type, owner_type)
-
-    @data('User', 'Usergroup')
-    def test_update_owner_type(self, owner_type):
-        """@Test: Update a host and specify an ``owner_type``.
+        for owner_type in ('User', 'Usergroup'):
+            with self.subTest(owner_type):
+                if owner_type == 'Usergroup' and bz_bug_is_open(1203865):
+                    continue  # instead of skip for compatibility with nose
```
Comment (sthirugn) [inline tests/foreman/api/test_host.py]: how will this be reported at the end of test run? a fail or a skip?

**Comment by omaciel** [comment]
ACK

@Ichimonji10 can you explain the bit where instead of explicitly skipping a test if the conditional statements are met, you are using `continue`?

File: tests/foreman/api/test_host.py [hunk 2]
```diff
@@ -70,27 +63,29 @@ def test_create_owner_type(self, owner_type):
...
+                host = entities.Host()
+                host.create_missing()
+                host.owner_type = owner_type
+                host = host.create(create_missing=False)
```
Comment (sthirugn) [inline tests/foreman/api/test_host.py]: create_missing=False is used here but not in the update test.  Why?

**Comment by sthirugn** [comment]
ACK pending comments

**Comment by Ichimonji10** [comment]
tl;dr I'm ensuring greater test coverage.

Also, I've outlined the issues here already, except for the bit about `continue`: https://github.com/SatelliteQE/robottelo/issues/2769#issuecomment-140191909

Nose doesn't know how to deal with subtests. When a failure or skip is encountered within a `with self.subTest(...)` context manager, nose fails fails the *entire* test. In other words, nose just completely ignores the context manager. This is also true of py.test (and nose2, IIRC). Nose looks at this code:

```python
for item in items:
    with self.subTest(item):
        # test code here
```

And it acts as if this is what's happening:

```python
for item in items:
    # test code here
```

I'm using `continue` so that we get as much test coverage as possible given that our test runner is nose. Nose is still capable of running our tests - it's just not as capable as it should be.

There's a couple possible solutions:

* Use unittest2 as a test runner. (`python -m unittest2 ...`). This has the down-side that unittest2 isn't natively capable of producing JUnit test reports, and unittest-xml-reporting has some issues as discussed in https://github.com/SatelliteQE/robottelo/issues/2769#issuecomment-140191909
* Fix nose/py.test/nose2/whatever so it deals correctly with subtests.

**Comment by omaciel** [comment]
ACK part Deux

File: tests/foreman/api/test_host.py (hunk 2 above)
Comment (Ichimonji10) [inline tests/foreman/api/test_host.py]: Typically, `create_missing` only populates required attributes that have not yet been filled in. The one exception to this is `Host.create_missing`. It will only work if absolutely no attributes have been filled in. This is because creating missing attributes for a host is extremely complex, much more so than for any other entity. Our solution is to call `create_missing` first an *then* make some customizations.

The update test has an entirely different structure. It goes like this:

```
create a host  # create_missing called here
for owner_type in owner_types:
    update host's owner_type
```

So there's no point to work around the limitations of `Host.create_missing` in that test.

Comment (Ichimonji10) [inline tests/foreman/api/test_host.py]: See: http://nailgun.readthedocs.org/en/latest/api/nailgun.entities.html#nailgun.entities.Host.create_missing

Comment (sthirugn) [inline tests/foreman/api/test_host.py]: :+1:

**Comment by sthirugn** [comment]
@Ichimonji10 if I understand this right - for now (until we implement any of the proposed solutions) the test which `continue`s will be reported as PASS.

**Comment by omaciel** [comment]
It will fail and not be skipped

**Comment by Ichimonji10** [comment]
> the test which `continue`s will be reported as PASS.

@sthirugn Yes, you're correct. Tests which use `continue` will PASS.

By the way, this counts as three tests:

```python
@data(1, 2, 3):
def test_greater_than_zero(self, number):
    self.assertGreater(number, 0)
```

But this only counts as one test:

```python
def test_greater_than_zero(self):
    for number in (1, 2, 3):
        with self.subTest(number):
            self.assertGreater(number, 0)
```

**Comment by sthirugn** [comment]
@Ichimonji10 gotcha.  Now I am curious to 
* see how jenkins (nosetests) will report these results 
* see how this brings down our our total test case numbers :)
